File: src/main/java/day04/UDFOperator.java
package day04;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.sql.DataFrameReader;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SQLContext;
import org.apache.spark.sql.api.java.UDF1;
import org.apache.spark.sql.types.DataTypes;

public class UDFOperator {
    public static void main(String[] args){

        SparkConf conf = new SparkConf().setAppName("UDFOperator").setMaster("local");

        JavaSparkContext sc = new JavaSparkContext(conf);

        SQLContext sqlContext = new SQLContext(sc);

        DataFrameReader reader = sqlContext.read().format("jdbc");

        reader.option("url","jdbc:mysql://localhost:3306/sakila?useSSL=false");
        reader.option("dbtable","actor");
        reader.option("driver","com.mysql.jdbc.Driver");
        reader.option("user","root");
        reader.option("password","root");

        Dataset<Row> load = reader.load();

        load.registerTempTable("actor");

        sqlContext.udf().register("strLen", new UDF1<String, Integer>() {

            public Integer call(String s) throws Exception {
                return s.length();
            }
        }, DataTypes.IntegerType);

        Dataset<Row> sql = sqlContext.sql("select first_name,strLen(first_name) from actor where actor_id<20");

        sql.show();


    }
}

File: src/main/scala/day18/UDFOperator.scala
package day18

import org.apache.spark.SparkConf
import org.apache.spark.sql.SparkSession

object UDFOperator {
  def main(args: Array[String]): Unit = {
    val conf = new SparkConf().setAppName("JDBCDataSource").setMaster("local")

    //val sc = new SparkContext(conf)

    val session = SparkSession.builder().config(conf).getOrCreate()

    //定义mysql信息
    val jdbcDF = session.read.format("jdbc").options(
      Map("url"->"jdbc:mysql://localhost:3306/sakila?useSSL=false",
        "dbtable"->"actor",
        "driver"->"com.mysql.jdbc.Driver",
        "user"-> "root",
        "password"->"root")).load()

    jdbcDF.createTempView("actor")
    //udf user defined function自定义聚合函数
    session.udf.register("strLen",(str:String)=>str.length())
    //session.sql("select first_name,strLen(first_name) from actor where actor_id<20").show()
    //udaf  user defined Aggregate function自定义聚合函数
    session.udf.register("strCount",new StringCount)
    session.sql("select first_name,strCount(first_name) from actor group by first_name").show()


  }
}
